modules/api/pkg/ee/clusterbackup/backupstoragelocation/handler.go

```diff
@@ -0,0 +1,458 @@
+			existingSecret = false
+			dataBytes, err := backupProvider.GetStorageLocationCreds(ctx, secretRefName)
+			if err != nil {
+				return nil, fmt.Errorf("please provide correct creds, error: %v", err.Error())
+			}
+			if dataBytes["accessKeyId"] == nil || dataBytes["secretAccessKey"] == nil {
+				return nil, fmt.Errorf("bsl credentials secret is not set correctly")
+			}
+			cloudCreds, err := getVeleroCloudCredentials(dataBytes["accessKeyId"], dataBytes["secretAccessKey"])
+			if err != nil {
+				return nil, fmt.Errorf("failed to generate Velero cloud-credentials template: %w", err)
+			}
+			if bsl.Spec.Credential.Key == "" {
+				bsl.Spec.Credential.Key = defaultCloudCredentialsSecretKeyName
+			}
+			secret = &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secretRefName,
+					Namespace: clusterbackup.UserClusterBackupNamespace,
+					Labels:    getBSLLabels(bslFullName, projectID, clusterID),
+				},
+				Data: map[string][]byte{
+					bsl.Spec.Credential.Key: cloudCreds,
+				},
+			}
+		}
+	}
+
+	bsl.ObjectMeta = metav1.ObjectMeta{
+		Name:      bslFullName,
+		Namespace: clusterbackup.UserClusterBackupNamespace,
+		Labels:    getBSLLabels(bslName, projectID, clusterID),
+	}
+
+	if !existingSecret {
+		bsl.Spec.Credential = &corev1.SecretKeySelector{
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: secret.Name,
+			},
+			Key: credentialsSecretKeyName,
+		}
+	}
+
+	if err := client.Create(ctx, bsl); err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+
+	ownerReferences := metav1.OwnerReference{
+		APIVersion: velerov1.SchemeGroupVersion.String(),
+		Kind:       bslKind,
+		Name:       bslFullName,
+		UID:        bsl.UID,
+	}
+	if !existingSecret {
+		secret.ObjectMeta.OwnerReferences = append(secret.ObjectMeta.OwnerReferences, ownerReferences)
+		if err := client.Create(ctx, secret); err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+	} else {
+		ownerRefExist := false
+		for _, ownerRef := range secret.ObjectMeta.OwnerReferences {
+			if ownerRef.UID == bsl.UID {
+				ownerRefExist = true
+				break
+			}
+		}
+		if !ownerRefExist {
+			updatedSecret := secret.DeepCopy()
+			updatedSecret.ObjectMeta.OwnerReferences = append(updatedSecret.ObjectMeta.OwnerReferences, ownerReferences)
+			if err := client.Patch(ctx, updatedSecret, ctrlruntimeclient.MergeFrom(secret)); err != nil {
+				return nil, common.KubernetesErrorToHTTPError(err)
+			}
+		}
+	}
+	return bsl, nil
+}
+
+func createHashFromBslSpec(bslSpec velerov1.BackupStorageLocationSpec) string {
+	s3Url := bslSpec.Config[bslS3Url]
+	region := bslSpec.Config[bslRegion]
+	hash := sha256.Sum256([]byte(s3Url + region + bslSpec.ObjectStorage.Bucket + bslSpec.ObjectStorage.Prefix))
+	hashStr := hex.EncodeToString(hash[:])[:12]
+	hashStr = strings.ToLower(hashStr)
+	hashStr = strings.ReplaceAll(hashStr, "_", "-")
+	if len(hashStr) > 6 {
+		hashStr = hashStr[:6]
+	}
+	return hashStr
+}
```

Comment:
Can we improve this somehow? For example, use a random string of 6 characters and use labels to store information regarding the actual BSL name/namespace etc.